### DISCUSSION TITLE
Would it be possible to pass a content:// URI or byte stream to be uploaded?

### DISCUSSION THREAD

**Discussion by Ansis100** [781flyingdutchman/background_downloader#398] [Q&A]
If you want to upload a file that was provided to you by the Storage Access Framework, you receive a content:// URI, which [currently cannot be handled by Dart.](https://github.com/flutter/flutter/issues/147037) The current accepted solution is to cache the file in the application directory and then upload it. This is not a great solution because devices with low storage space will not be able to upload files.

The content:// URIs are unsupported by **Dart** which means *theoretically* you could pass a URI to the `enqueue` or `upload` methods (instead of `directory` and `filename`) and do the actual file reading in Android-specific code, without caching. Alternatively, you could use a package like [saf_stream](https://pub.dev/packages/saf_stream), which provides a readable byte stream of a SAF file and pass that to the upload methods.

**Comment by 781flyingdutchman** [comment]
That's an interesting idea, and seems feasible.  You would set the `directory` to `.root` and the `filename` to the uri string. It would throw on platforms other than Android upon task creation (and only work for `UploadTask` to begin with). Byte streams are not a good idea, as they don't work when the app gets suspended.

I'm labeling this as an enhancement - it will take me time to get to this though.

**Reply by Ansis100** [reply]
If you have other priorities, I could eventually start some work on this.
I don't understand the reasoning behind setting `directory` to `.root`, though. Is `.root` some identifier I'm not familiar with?

**Reply by 781flyingdutchman** [reply]
I actually think it's not that hard so can probably get it done soonish, just didn't want to delay today's release. 
Ignore the `.root` point, as I'll probably just create a named constructor for this, just like there is one for `forFile`

**Reply by 781flyingdutchman** [reply]
Implemented in 8.7.0
